catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/QueryResultIterable.java

```diff
@@ -0,0 +1,111 @@
+
+import org.apache.commons.lang.Validate;
+
+import ddf.catalog.CatalogFramework;
+import ddf.catalog.data.Result;
+import ddf.catalog.operation.Query;
+
+/**
+ * Class that facilitates iteration over individual catalog results.
+ * <br/>
+ * Throws a {@link CatalogQueryException} if anything goes wrong during iteration or
+ * querying.
+ */
+public class QueryResultIterable implements Iterable<Result> {
+    private final CatalogFramework catalog;
+
+    private final Query query;
+
+    /**
+     * Constructor for QueryResultIterable.
+     *
+     * @param catalogFramework catalog to query
+     * @param query            query used to query the catalog framework for results. Query
+     *                         parameters such as {@link Query#getPageSize()} and
+     *                         {@link Query#getStartIndex()} will be used and are guaranteed to be
+     *                         respected when valid, regardless of any limit imposed by the catalog
+     *                         framework on the result size.
+     */
+    public QueryResultIterable(CatalogFramework catalogFramework, Query query) {
+        Validate.notNull(catalogFramework, "Catalog is null");
+        Validate.notNull(query, "queryRequest is null");
+
+        this.catalog = catalogFramework;
+        this.query = query;
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * The {@link Iterator} will retrieve results from the {@link CatalogFramework} as needed,
+     * based on the start index and page size provided in the query.
+     */
+    @Override
+    public Iterator<Result> iterator() {
+        return new QueryResultIterator(new QueryResultPaginator(catalog, query));
+    }
+
+    @Override
+    public Spliterator<Result> spliterator() {
+        int characteristics = Spliterator.DISTINCT;
+        return Spliterators.spliteratorUnknownSize(this.iterator(), characteristics);
+    }
+
+    static class QueryResultIterator implements Iterator<Result> {
+
+        private QueryResultPaginator queryResultPaginator;
+
+        private Iterator<Result> queriedResults = null;
+
+        public QueryResultIterator(QueryResultPaginator queryResultPaginator) {
+            this.queryResultPaginator = queryResultPaginator;
+        }
+
+        @Override
+        public boolean hasNext() {
+
+            if (isQueriedResultListNullOrEmpty()) {
+                return queryResultPaginator.hasNext();
+            }
+
+            return true;
+        }
+
+        @Override
+        public Result next() {
+
+            if (isQueriedResultListNullOrEmpty()) {
+                queriedResults = queryResultPaginator.next()
+                        .iterator();
+            }
+
+            return queriedResults.next();
+        }
+
+        private boolean isQueriedResultListNullOrEmpty() {
+            return queriedResults == null || !queriedResults.hasNext();
+        }
+
+        void setQueriedResults(Iterator<Result> queriedResults) {
```

Comment:
❗️ We shouldn't need this method. Everything should be tested through the paginator mock and public API.